vcd/remove_leftovers_test.go

```diff
@@ -291,6 +301,195 @@ func removeLeftovers(govcdClient *govcd.VCDClient, verbose bool) error {
+
+			if err != nil {
+				return fmt.Errorf("error retrieving NSX-T ALB Service Engine Groups list: %s", err)
+			}
+			for _, albServiceEngineGroup := range albServiceEngineGroups {
+				albServiceEngineGroupToBeDeleted := shouldDeleteEntity(alsoDelete, doNotDelete, albServiceEngineGroup.NsxtAlbServiceEngineGroup.Name, "vcd_nsxt_alb_service_engine_group", 2, verbose)
+				if albServiceEngineGroupToBeDeleted {
+
+					// --------------------------------------------------------------
+					// NSX-T ALB Tenant Configuration cleanup
+					// --------------------------------------------------------------
+					err = removeLeftoversNsxtAlbTenant(govcdClient, verbose)
+					if err != nil {
+						return fmt.Errorf("error removing NSX-T ALB Tenant leftovers: %s", err)
+					}
+
+					// vcd_nsxt_alb_service_engine_group deletion
+					err = albServiceEngineGroup.Delete()
+					if err != nil {
+						return fmt.Errorf("error deleting NSX-T ALB Service Engine Group '%s': %s", albServiceEngineGroup.NsxtAlbServiceEngineGroup.Name, err)
+					}
+				}
+			}
+
+			// vcd_nsxt_alb_cloud deletion
+			if albCloudToBeDeleted {
+				err = albCloud.Delete()
+				if err != nil {
+					return fmt.Errorf("error deleting NSX-T ALB Cloud '%s': %s", albCloud.NsxtAlbCloud.Name, err)
+				}
+			}
+		}
+
+		// vcd_nsxt_alb_controller deletion
+		if albControllertoBeDeleted {
+			err = albController.Delete()
+			if err != nil {
+				return fmt.Errorf("error deleting NSX-T ALB Controller '%s': %s", albController.NsxtAlbController.Name, err)
+			}
+		}
+	}
+	return nil
+}
+
+// removeLeftoversNsxtAlbTenant is responsible for cleanup up NSX-T ALB leftovers in Edge Gateways
+// (tenant configuration)
+// 1. Edge Gateway settings (vcd_nsxt_alb_settings)
+// 2. Edge Gateway Service Engine Group assignment (vcd_nsxt_alb_edgegateway_service_engine_group)
+// 3. NSX-T ALB Pools (vcd_nsxt_alb_pool)
+// 4. NSX-T ALB Virtual Services (vcd_nsxt_alb_virtual_service)
+func removeLeftoversNsxtAlbTenant(govcdClient *govcd.VCDClient, verbose bool) error {
+	// Iterate over Edge Gateways and ensure no tenant ALB configurations remain
+	edgeGateways, err := govcdClient.GetAllNsxtEdgeGateways(nil)
+	if err != nil {
+		return fmt.Errorf("error retrieving NSX-T Edge Gateway list: %s", err)
+	}
+
+	for _, edgeGateway := range edgeGateways {
+
+		// --------------------------------------------------------------
+		// vcd_nsxt_alb_settings
+		// --------------------------------------------------------------
+		albSetting, err := edgeGateway.GetAlbSettings()
+		if err != nil {
+			return fmt.Errorf("error retrieving NSX-T ALB Edge Gateway settings: %s", err)
+		}
+
+		if albSetting.Enabled {
+			// Real name of Edge Gateway would not be matched, but ALB needs to be deleted therefore adding a prefix Test
+			edgeNameWithPrefix := "Test" + edgeGateway.EdgeGateway.Name
+			toBeDisabledAlbSettings := shouldDeleteEntity(alsoDelete, doNotDelete, edgeNameWithPrefix, "vcd_nsxt_alb_settings", 3, verbose)
```

Comment:
As it matches the `isTest` regexp (we force to have the "Test" prefix), and the resources created by tests are not in the `doNotDelete` slice, wouldn't this always be `true`?

Reply:
Yes - it will.
Due to the nature of ALB we have all of the resources always match now (we can't keep the controller we need to test we can create it).

The only reason why I tried to follow the pattern for all ALB components is if for whatever reason we decide to keep it (maybe during the development phase).

The ALB Settings itself is even uglier because it is not a separate entity, but only a setting of Edge Gateway

Reply:
Got it, thanks!